File: apps/tauri/src/lib/vpn/privacy-leak-probe.ts
import { probeExitIp } from './exit-ip-probe'
import { detectLocalIpv6Desktop } from './privacy-bridge'

export interface PrivacyLeakProbeResult {
  exitIp: string | null
  exitIpLooksProtected: boolean
  ipv6LocalActive: boolean
  dnsReachable: boolean
  passed: boolean
}

const BASELINE_IP_KEY = 'tauri_privacy_baseline_ip'
const DNS_PROBE_URL = 'https://cloudflare-dns.com/dns-query?name=example.com&type=A'

export function savePrivacyBaselineIp(ip: string | null) {
  if (ip) localStorage.setItem(BASELINE_IP_KEY, ip)
}

export function loadPrivacyBaselineIp(): string | null {
  return localStorage.getItem(BASELINE_IP_KEY)
}

/** 纯逻辑评估，便于单测。 */
export function evaluatePrivacyLeakProbe(input: {
  exitIp: string | null
  baselineIp?: string | null
  ipv6LocalActive: boolean
  dnsReachable: boolean
  ipv6ProtectionEnabled?: boolean
}): PrivacyLeakProbeResult {
  const ipv6ProtectionEnabled = input.ipv6ProtectionEnabled ?? true
  const ipv6Risk = ipv6ProtectionEnabled && input.ipv6LocalActive
  const protectedIp =
    !!input.exitIp &&
    (!input.baselineIp || input.exitIp.toLowerCase() !== input.baselineIp.toLowerCase())
  const passed = protectedIp && !ipv6Risk && input.dnsReachable
  return {
    exitIp: input.exitIp,
    exitIpLooksProtected: protectedIp,
    ipv6LocalActive: input.ipv6LocalActive,
    dnsReachable: input.dnsReachable,
    passed,
  }
}

async function probeDns(): Promise<boolean> {
  try {
    const res = await fetch(DNS_PROBE_URL, {
      method: 'GET',
      headers: { Accept: 'application/dns-json' },
      cache: 'no-store',
    })
    return res.status >= 200 && res.status < 300
  } catch {
    return false
  }
}

/** 桌面端轻量隐私泄露自检（出口 IP / DNS / IPv6）。 */
export async function runPrivacyLeakProbe(options?: {
  baselineIp?: string | null
  ipv6ProtectionEnabled?: boolean
}): Promise<PrivacyLeakProbeResult> {
  const baseline = options?.baselineIp ?? loadPrivacyBaselineIp()
  const exit = await probeExitIp().catch(() => null)
  const dnsOk = await probeDns()
  const ipv6Local = await detectLocalIpv6Desktop()
  return evaluatePrivacyLeakProbe({
    exitIp: exit?.ip ?? null,
    baselineIp: baseline,
    ipv6LocalActive: ipv6Local,
    dnsReachable: dnsOk,
    ipv6ProtectionEnabled: options?.ipv6ProtectionEnabled ?? true,
  })
}

export function formatPrivacyProbeMessage(result: PrivacyLeakProbeResult): string {
  if (result.passed) {
    return `自检通过：出口 IP ${result.exitIp ?? '-'}`
  }
  const parts = ['自检未完全通过']
  if (!result.exitIpLooksProtected) parts.push('出口 IP 异常')
  if (result.ipv6LocalActive) parts.push('IPv6 风险')
  if (!result.dnsReachable) parts.push('DNS 异常')
  return parts.join(' · ')
}

File: apps/tauri/platforms/ios/KuayunVPN/Core/VPN/PrivacyLeakProbe.swift
import Foundation

struct ExitIpInfo {
    let ip: String
    let country: String?
    let region: String?
    let city: String?
}

enum ExitIpProbe {
    private static let geoURL = URL(string: "https://ip-api.com/json/?fields=status,query,country,regionName,city")!

    static func probe(timeout: TimeInterval = 8) async -> ExitIpInfo? {
        var request = URLRequest(url: geoURL)
        request.httpMethod = "GET"
        request.cachePolicy = .reloadIgnoringLocalCacheData
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
            struct Payload: Decodable {
                let status: String?
                let query: String?
                let country: String?
                let regionName: String?
                let city: String?
            }
            let json = try JSONDecoder().decode(Payload.self, from: data)
            guard json.status == "success", let ip = json.query?.trimmingCharacters(in: .whitespacesAndNewlines), !ip.isEmpty else {
                return nil
            }
            return ExitIpInfo(ip: ip, country: json.country, region: json.regionName, city: json.city)
        } catch {
            return nil
        }
    }
}

struct PrivacyLeakProbeResult {
    let exitIp: String?
    let exitIpLooksProtected: Bool
    let ipv6LocalActive: Bool
    let dnsReachable: Bool
    let passed: Bool
}

enum PrivacyLeakProbe {
    private static let baselineKey = "kuayun_ios_privacy_baseline_ip"
    private static let dnsURL = URL(string: "https://cloudflare-dns.com/dns-query?name=example.com&type=A")!

    static func saveBaselineIp(_ ip: String?) {
        guard let ip, !ip.isEmpty else { return }
        UserDefaults.standard.set(ip, forKey: baselineKey)
    }

    static func loadBaselineIp() -> String? {
        UserDefaults.standard.string(forKey: baselineKey)
    }

    static func evaluate(
        exitIp: String?,
        baselineIp: String?,
        ipv6LocalActive: Bool,
        dnsReachable: Bool
    ) -> PrivacyLeakProbeResult {
        let protected = exitIp != nil && (baselineIp == nil || exitIp!.lowercased() != baselineIp!.lowercased())
        let passed = protected && !ipv6LocalActive && dnsReachable
        return PrivacyLeakProbeResult(
            exitIp: exitIp,
            exitIpLooksProtected: protected,
            ipv6LocalActive: ipv6LocalActive,
            dnsReachable: dnsReachable,
            passed: passed
        )
    }

    static func run() async -> PrivacyLeakProbeResult {
        async let exit = ExitIpProbe.probe()
        async let dns = probeDns()
        let baseline = loadBaselineIp()
        let exitInfo = await exit
        let dnsOk = await dns
        // iOS 无 Rust IPv6 探测；NE 全隧道下通常无本地 IPv6 泄露，保守标 false
        return evaluate(
            exitIp: exitInfo?.ip,
            baselineIp: baseline,
            ipv6LocalActive: false,
            dnsReachable: dnsOk
        )
    }

    static func formatMessage(_ result: PrivacyLeakProbeResult) -> String {
        if result.passed { return "自检通过：出口 IP \(result.exitIp ?? "-")" }
        var parts = ["自检未完全通过"]
        if !result.exitIpLooksProtected { parts.append("出口 IP 异常") }
        if result.ipv6LocalActive { parts.append("IPv6 风险") }
        if !result.dnsReachable { parts.append("DNS 异常") }
        return parts.joined(separator: " · ")
    }

    private static func probeDns() async -> Bool {
        var request = URLRequest(url: dnsURL)
        request.setValue("application/dns-json", forHTTPHeaderField: "Accept")
        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse else { return false }
            return (200..<300).contains(http.statusCode)
        } catch {
            return false
        }
    }
}
